data/src/main/java/com/linkedin/data/schema/annotation/AnnotationCheckResolvedPropertiesVisitor.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (c) 2020 LinkedIn Corp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.linkedin.data.schema.annotation;
+
+import com.linkedin.data.schema.DataSchema;
+import com.linkedin.data.schema.DataSchemaTraverse;
+import com.linkedin.data.schema.PathSpec;
+import com.linkedin.data.schema.RecordDataSchema;
+import com.linkedin.data.schema.UnionDataSchema;
+import com.linkedin.data.schema.annotation.SchemaAnnotationHandler.AnnotationCheckContext;
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+import org.apache.commons.lang3.tuple.Pair;
+
+
+/**
+ * This visitor is used to get node in schema to it's resolvedProperties.
+ * The nodesToResolvedPropertiesMap will be used for schema annotation compatibility check
+ *
+ * @author Yingjie Bi
+ */
+public class AnnotationCheckResolvedPropertiesVisitor implements SchemaVisitor
+{
+  private Map<PathSpec, Pair<AnnotationCheckContext, Map<String, Object>>> _nodeToResolvedPropertiesMap = new HashMap<>();
+
+  @Override
+  public void callbackOnContext(TraverserContext context, DataSchemaTraverse.Order order)
+  {
+    // Skip execute when order is POST_ORDER to avoid traverse same node twice.
+    // Only execute this callback when order is PRE_ORDER.
+    if (order == DataSchemaTraverse.Order.POST_ORDER)
+    {
+      return;
+    }
+
+    DataSchema currentSchema = context.getCurrentSchema();
+    Map<String, Object> resolvedProperties = currentSchema.getResolvedProperties();
+    if (resolvedProperties == null || resolvedProperties.isEmpty())
```

Comment:
Leave the empty resolvedProperties for fields without any annotation. It would simplify the compatibility checking logic.

Reply:
Yes, that's right. 
Updated.